File: mobile/android/app/src/main/java/com/bitunnel/mobile/mux/MuxSession.kt
package com.bitunnel.mobile.mux

import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.concurrent.BlockingQueue
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import kotlin.concurrent.thread

data class MuxFrame(
    val type: Int,
    val channelId: Long,
    val payload: ByteArray
)

class MuxSession(
    private val inputStream: InputStream,
    private val outputStream: OutputStream,
    password: String?
) {
    companion object {
        const val TYPE_DATA = 1
        const val TYPE_CREATE = 2
        const val TYPE_CLOSE = 3
        const val TYPE_AUTH = 4
        const val TYPE_AUTH_RES = 5
        const val TYPE_CREATE_ACK = 6

        private const val AES_ALG = "AES/GCM/NoPadding"
        private const val AES_IV_LEN = 12
        private const val AES_TAG_LEN = 16
        private const val HEADER_SIZE = 9
        private const val MAX_FRAME_SIZE = 16 * 1024 * 1024
    }

    private val encryptionKey: ByteArray? =
        if (!password.isNullOrEmpty()) {
            val md = MessageDigest.getInstance("SHA-256")
            md.digest(password.toByteArray())
        } else null

    @Volatile
    var isAuthenticated = false
        private set

    @Volatile
    private var closed = false

    private val frameListeners = ConcurrentHashMap<Long, BlockingQueue<MuxFrame>>()
    private var readerThread: Thread? = null
    private var controlHandler: ((MuxFrame) -> Unit)? = null

    fun startReader(handler: (MuxFrame) -> Unit) {
        controlHandler = handler
        readerThread = thread(isDaemon = false, name = "mux-reader") {
            try {
                while (!closed) {
                    val frame = readFrame() ?: break
                    val queue = frameListeners[frame.channelId]
                    if (queue != null) {
                        queue.put(frame)
                    } else {
                        controlHandler?.invoke(frame)
                    }
                }
            } catch (_: Exception) {
            } finally {
                closed = true
            }
        }
    }

    fun stopReader() {
        closed = true
        readerThread?.join(2000)
        readerThread = null
        frameListeners.clear()
    }

    fun subscribeChannel(channelId: Long): BlockingQueue<MuxFrame> {
        val queue = LinkedBlockingQueue<MuxFrame>()
        frameListeners[channelId] = queue
        return queue
    }

    fun unsubscribeChannel(channelId: Long) {
        frameListeners.remove(channelId)
    }

    fun sendAuth(password: String, clientId: String) {
        val payload = """{"password":"$password","clientId":"$clientId"}""".toByteArray()
        sendFrame(TYPE_AUTH, 0, payload)
    }

    fun sendAuthRes(ok: Boolean) {
        sendFrame(TYPE_AUTH_RES, 0, byteArrayOf(if (ok) 1 else 0))
        if (ok) isAuthenticated = true
    }

    fun sendData(channelId: Long, data: ByteArray) {
        sendFrame(TYPE_DATA, channelId, data)
    }

    fun sendCreate(channelId: Long, host: String, port: Int) {
        val meta = """{"type":"forward","host":"$host","port":$port}""".toByteArray()
        sendFrame(TYPE_CREATE, channelId, meta)
    }

    fun sendCreateAck(channelId: Long, success: Boolean) {
        sendFrame(TYPE_CREATE_ACK, channelId, byteArrayOf(if (success) 1 else 0))
    }

    fun sendClose(channelId: Long) {
        sendFrame(TYPE_CLOSE, channelId, ByteArray(0))
    }

    @Synchronized
    fun sendFrame(type: Int, channelId: Long, payload: ByteArray) {
        if (closed || encryptionKey == null && type != TYPE_AUTH && type != TYPE_AUTH_RES) return
        val encrypted = encrypt(payload)
        val header = ByteBuffer.allocate(HEADER_SIZE)
        header.put(type.toByte())
        header.putInt(channelId.toInt())
        header.putInt(encrypted.size)
        try {
            outputStream.write(header.array())
            outputStream.write(encrypted)
            outputStream.flush()
        } catch (e: Exception) {
            closed = true
            throw e
        }
    }

    fun readFrame(): MuxFrame? {
        try {
            val header = ByteArray(HEADER_SIZE)
            var offset = 0
            while (offset < HEADER_SIZE) {
                val n = inputStream.read(header, offset, HEADER_SIZE - offset)
                if (n < 0) return null
                offset += n
            }
            val buf = ByteBuffer.wrap(header)
            val type = buf.get().toInt() and 0xFF
            val channelId = buf.getInt().toLong() and 0xFFFFFFFFL
            val payloadLen = buf.getInt()

            if (payloadLen < 0 || payloadLen > MAX_FRAME_SIZE + AES_IV_LEN + AES_TAG_LEN) {
                return null
            }

            val encrypted = ByteArray(payloadLen)
            offset = 0
            while (offset < payloadLen) {
                val n = inputStream.read(encrypted, offset, payloadLen - offset)
                if (n < 0) return null
                offset += n
            }

            val payload = decrypt(encrypted)
            return MuxFrame(type, channelId, payload)
        } catch (e: Exception) {
            return null
        }
    }

    private fun encrypt(plaintext: ByteArray): ByteArray {
        if (encryptionKey == null) return plaintext
        val cipher = Cipher.getInstance(AES_ALG)
        val iv = ByteArray(AES_IV_LEN)
        SecureRandom().nextBytes(iv)
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(encryptionKey, "AES"), GCMParameterSpec(AES_TAG_LEN * 8, iv))
        val encrypted = cipher.doFinal(plaintext)
        return iv + encrypted
    }

    private fun decrypt(data: ByteArray): ByteArray {
        if (encryptionKey == null) return data
        if (data.size < AES_IV_LEN + AES_TAG_LEN) return ByteArray(0)
        val cipher = Cipher.getInstance(AES_ALG)
        val iv = data.copyOfRange(0, AES_IV_LEN)
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(encryptionKey, "AES"), GCMParameterSpec(AES_TAG_LEN * 8, iv))
        return cipher.doFinal(data, AES_IV_LEN, data.size - AES_IV_LEN)
    }

    fun close() {
        closed = true
        try { inputStream.close() } catch (_: Exception) {}
        try { outputStream.close() } catch (_: Exception) {}
    }

    val isClosed: Boolean get() = closed
}

File: mobile/ios/Runner/MuxSession.swift
import Foundation
import Network
import CryptoKit

struct MuxFrame {
    let type: Int
    let channelId: UInt32
    let payload: Data
}

class MuxSession {
    static let TYPE_DATA = 1
    static let TYPE_CREATE = 2
    static let TYPE_CLOSE = 3
    static let TYPE_AUTH = 4
    static let TYPE_AUTH_RES = 5
    static let TYPE_CREATE_ACK = 6

    private let HEADER_SIZE = 9
    private let AES_IV_LEN = 12
    private let AES_TAG_LEN = 16
    private let MAX_FRAME_SIZE = 16 * 1024 * 1024

    private let connection: NWConnection
    private let encryptionKey: SymmetricKey?
    private let readQueue = DispatchQueue(label: "mux-reader")
    private let writeLock = NSLock()

    private var frameQueues: [UInt32: BlockingQueue<MuxFrame>] = [:]
    private var frameQueuesLock = NSLock()

    var isAuthenticated = false
    private(set) var isClosed = false
    var controlHandler: ((MuxFrame) -> Void)?

    init(connection: NWConnection, password: String?) {
        self.connection = connection
        if let pwd = password, !pwd.isEmpty {
            let hash = SHA256.hash(data: pwd.data(using: .utf8)!)
            encryptionKey = SymmetricKey(data: hash)
        } else {
            encryptionKey = nil
        }
    }

    func startReader(handler: @escaping (MuxFrame) -> Void) {
        controlHandler = handler
        readQueue.async { [weak self] in
            self?.readLoop()
        }
    }

    func stopReader() {
        isClosed = true
    }

    func subscribeChannel(_ channelId: UInt32) -> BlockingQueue<MuxFrame> {
        let q = BlockingQueue<MuxFrame>()
        frameQueuesLock.lock()
        frameQueues[channelId] = q
        frameQueuesLock.unlock()
        return q
    }

    func unsubscribeChannel(_ channelId: UInt32) {
        frameQueuesLock.lock()
        frameQueues.removeValue(forKey: channelId)
        frameQueuesLock.unlock()
    }

    func sendAuth(password: String, clientId: String) {
        let payload = "{\"password\":\"\(password)\",\"clientId\":\"\(clientId)\"}"
        sendFrame(type: MuxSession.TYPE_AUTH, channelId: 0, payload: payload.data(using: .utf8) ?? Data())
    }

    func sendAuthRes(_ ok: Bool) {
        sendFrame(type: MuxSession.TYPE_AUTH_RES, channelId: 0, payload: Data([ok ? 1 : 0]))
        if ok { isAuthenticated = true }
    }

    func sendData(channelId: UInt32, data: Data) {
        sendFrame(type: MuxSession.TYPE_DATA, channelId: channelId, payload: data)
    }

    func sendCreate(channelId: UInt32, host: String, port: Int) {
        let meta = "{\"type\":\"forward\",\"host\":\"\(host)\",\"port\":\(port)}"
        sendFrame(type: MuxSession.TYPE_CREATE, channelId: channelId, payload: meta.data(using: .utf8) ?? Data())
    }

    func sendCreateAck(channelId: UInt32, success: Bool) {
        sendFrame(type: MuxSession.TYPE_CREATE_ACK, channelId: channelId, payload: Data([success ? 1 : 0]))
    }

    func sendClose(channelId: UInt32) {
        sendFrame(type: MuxSession.TYPE_CLOSE, channelId: channelId, payload: Data())
    }

    func close() {
        isClosed = true
        connection.cancel()
    }

    // MARK: - Frame I/O

    private func sendFrame(type: Int, channelId: UInt32, payload: Data) {
        guard !isClosed else { return }
        let encrypted = encrypt(payload)
        var header = Data(capacity: HEADER_SIZE + encrypted.count)
        header.append(UInt8(type))
        header.append(contentsOf: withUnsafeBytes(of: channelId.bigEndian) { Data($0) })
        header.append(contentsOf: withUnsafeBytes(of: UInt32(encrypted.count).bigEndian) { Data($0) })
        header.append(encrypted)

        writeLock.lock()
        connection.send(content: header, completion: .contentProcessed { _ in })
        writeLock.unlock()
    }

    private func readLoop() {
        while !isClosed {
            guard let frame = readFrameSync() else {
                isClosed = true
                return
            }
            var queue: BlockingQueue<MuxFrame>?
            frameQueuesLock.lock()
            queue = frameQueues[frame.channelId]
            frameQueuesLock.unlock()

            if let q = queue {
                q.put(frame)
            } else {
                controlHandler?(frame)
            }
        }
    }

    private func readFrameSync() -> MuxFrame? {
        guard let header = readBytesSync(length: HEADER_SIZE) else { return nil }
        guard header.count == HEADER_SIZE else { return nil }

        let type = Int(header[0])
        let channelId = header.withUnsafeBytes { $0.loadUnaligned(fromByteOffset: 1, as: UInt32.self) }.bigEndian
        let payloadLen = header.withUnsafeBytes { $0.loadUnaligned(fromByteOffset: 5, as: UInt32.self) }.bigEndian

        guard payloadLen <= MAX_FRAME_SIZE + AES_IV_LEN + AES_TAG_LEN else { return nil }
        guard payloadLen > 0 else { return MuxFrame(type: type, channelId: channelId, payload: Data()) }

        guard let encrypted = readBytesSync(length: Int(payloadLen)) else { return nil }
        let payload = decrypt(encrypted)
        return MuxFrame(type: type, channelId: channelId, payload: payload)
    }

    func readFrame() -> MuxFrame? {
        return readFrameSync()
    }

    private func readBytesSync(length: Int) -> Data? {
        var result = Data()
        let semaphore = DispatchSemaphore(value: 0)

        var lastError: NWError?
        var lastData: Data?

        connection.receive(minimumIncompleteLength: length, maximumLength: length) { data, _, _, error in
            lastData = data
            lastError = error
            semaphore.signal()
        }

        semaphore.wait()

        if let err = lastError {
            print("[MuxSession] read error: \(err)")
            return nil
        }
        guard let d = lastData, d.count == length else {
            return nil
        }
        return d
    }

    // MARK: - Encryption

    private func encrypt(_ plaintext: Data) -> Data {
        guard let key = encryptionKey else { return plaintext }
        do {
            let sealed = try AES.GCM.seal(plaintext, using: key)
            return sealed.nonce + sealed.ciphertext + sealed.tag
        } catch {
            return plaintext
        }
    }

    private func decrypt(_ data: Data) -> Data {
        guard let key = encryptionKey else { return data }
        guard data.count > AES_IV_LEN + AES_TAG_LEN else { return Data() }
        do {
            let nonce = try AES.GCM.Nonce(data: data.prefix(AES_IV_LEN))
            let box = try AES.GCM.SealedBox(
                nonce: nonce,
                ciphertext: data.dropFirst(AES_IV_LEN).dropLast(AES_TAG_LEN),
                tag: data.suffix(AES_TAG_LEN)
            )
            return try AES.GCM.open(box, using: key)
        } catch {
            return data
        }
    }
}

// MARK: - Blocking Queue

class BlockingQueue<T> {
    private var items: [T] = []
    private let lock = NSLock()
    private let sem = DispatchSemaphore(value: 0)

    func put(_ item: T) {
        lock.lock()
        items.append(item)
        lock.unlock()
        sem.signal()
    }

    func poll(timeout: DispatchTime) -> T? {
        let result = sem.wait(timeout: timeout)
        if result == .timedOut { return nil }
        lock.lock()
        let item = items.removeFirst()
        lock.unlock()
        return item
    }

    func take() -> T {
        sem.wait()
        lock.lock()
        let item = items.removeFirst()
        lock.unlock()
        return item
    }
}
